src/app/clusters/general-commissioning-server/general-commissioning-server.cpp

```diff
@@ -45,104 +48,194 @@ using namespace chip::DeviceLayer;
+            callable();
+        }
+    }
+}
+
+#define CheckSuccess(expr, code, onSuccess, onFailure)                                                                             \
     do                                                                                                                             \
     {                                                                                                                              \
         if (!::chip::ChipError::IsSuccess(expr))                                                                                   \
         {                                                                                                                          \
             commandObj->AddStatus(commandPath, Protocols::InteractionModel::Status::code, #expr);                                  \
+            CallIfFunction(onFailure);                                                                                             \
             return true;                                                                                                           \
         }                                                                                                                          \
+        CallIfFunction(onSuccess);                                                                                                 \
     } while (false)
 
 namespace {
 
+template <typename Provider, typename T>
+CHIP_ERROR ReadIfSupported(Provider * const provider, CHIP_ERROR (Provider::*const nonConstGetter)(T &),
+                           AttributeValueEncoder & aEncoder)
+{
+    if (nullptr == provider)
+    {
+        return CHIP_ERROR_PERSISTED_STORAGE_FAILED;
+    }
+
+    T value;
+    CHIP_ERROR err = (provider->*nonConstGetter)(value);
+    if (err != CHIP_NO_ERROR)
+    {
+        return err;
+    }
+
+    return aEncoder.Encode(value);
+}
+
+template <typename Provider, typename T>
+CHIP_ERROR ReadIfSupported(Provider * const provider, CHIP_ERROR (Provider::*const getter)(T &) const,
+                           AttributeValueEncoder & aEncoder)
+{
+    // Removing the const qualifier from the getter function pointer because there are a handful of getter functions that are not
+    // correctly marked as const.
+    using NonConstGetter          = CHIP_ERROR (Provider::*)(T &);
+    NonConstGetter nonConstGetter = reinterpret_cast<NonConstGetter>(getter);
+    return ReadIfSupported(provider, nonConstGetter, aEncoder);
+}
+
 class GeneralCommissioningAttrAccess : public AttributeAccessInterface
 {
 public:
-    // Register for the GeneralCommissioning cluster on all endpoints.
     GeneralCommissioningAttrAccess() : AttributeAccessInterface(Optional<EndpointId>::Missing(), GeneralCommissioning::Id) {}
 
-    CHIP_ERROR Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder) override;
+    CHIP_ERROR Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder)
```

Comment:
per offline feedback, moving code makes it more difficult for reviewers to confirm that the changes are low risk